aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -0,0 +1,248 @@
+        let stake_pool = StakePool {
+            current_stake: 0,
+            active: Vector::empty(),
+            pending_active: Vector::empty(),
+            pending_inactive: Vector::empty(),
+            inactive: Vector::empty(),
+        };
+        move_to(account, ValidatorInfo {
+            stake_pool,
+            consensus_pubkey,
+            network_address,
+        });
+    }
+
+    /// Rotate the consensus key of the validator.
+    fun rotate_consensus_key(account: &signer, consensus_pubkey: vector<u8>) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+        borrow_global_mut<ValidatorInfo>(addr).consensus_pubkey = consensus_pubkey;
+        if (Vector::contains(&validator_set.validators, &addr)) {
+            on_new_epoch(validator_set);
+        };
+    }
+
+    /// Initialize validator set to the core resource account.
+    fun initialize_validator_set(account: &signer, minimum_stake: u64, maximum_stake: u64) {
+        SystemAddresses::assert_core_resource(account);
+        move_to(account, ValidatorSet {
+            consensus_scheme: 0,
+            minimum_stake,
+            maximum_stake,
+            validators: Vector::empty(),
+            last_update_time_secs: Timestamp::now_seconds(),
+        });
+    }
+
+    /// Initiate by the validator info owner
+    fun join_validator_set(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_info = borrow_global_mut<ValidatorInfo>(addr);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+        assert!(!Vector::contains(&validator_set.validators, &addr), 0);
+
+        assert!(validator_info.stake_pool.current_stake >= validator_set.minimum_stake, 0);
+        assert!(validator_info.stake_pool.current_stake <= validator_set.maximum_stake, 0);
+
+        on_new_epoch(validator_set);
+        Vector::push_back(&mut validator_set.validators, addr);
+    }
+
+    /// Initiate by the validator info owner.
+    fun leave_validator_set(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+
+        let (exist, index) = Vector::index_of(&validator_set.validators, &addr);
+        assert!(exist, 0);
+        on_new_epoch(validator_set);
+        Vector::swap_remove(&mut validator_set.validators, index);
+        assert!(Vector::length(&validator_set.validators) > 0, 0);
+    }
+
+    /// Triggers when validator set changes or after certain time period (in block prologue).
+    fun on_new_epoch(validator_set: &mut ValidatorSet) acquires ValidatorInfo {
+        let current_time = Timestamp::now_seconds();
+        assert!(validator_set.last_update_time_secs + MINIMUM_RECONFIG_PERIOD > current_time, 0);
```

Comment:
Did I read it right? You can only join or leave the validator set if someone else did it in the last hour? 

Or do you actually mean the opposite (validator set changes at most once a hour)? Then if people wants to join or flee the validator set or rotate key, they'd need to race, and waste gas. I guess this relates to the "buffer all the changes and make them effective at the epoch discussion".

Reply:
good catch!